lib/extensions/ar_miq_set.rb

```diff
@@ -37,38 +27,51 @@ def make_memberof(set)
   end
 end # module ActsAsMiqSetMember
 
+module ActiveRecord
+  class Base
+    def self.acts_as_miq_set_member
+      include ActsAsMiqSetMember
+    end
+
+    def self.acts_as_miq_set(model_class = nil)
+      include ActsAsMiqSet
+
+      self.model_class = model_class unless model_class.nil?
+    end
+  end
+end
+
 module ActsAsMiqSet
   extend ActiveSupport::Concern
   included do
-    include RelationshipMixin
-    self.default_relationship_type ||= "membership"
-
     include UuidMixin
 
     self.table_name         = "miq_sets"
     self.inheritance_column = :set_type
 
     serialize :set_data
 
-    validates_presence_of     :name
-    validates_uniqueness_of   :name,
-                              :scope => [:set_type, :userid, :group_id],
-                              :if    => proc { |c| c.class.in_my_region.exists?(:name => c.name) }
-    validates_presence_of     :description
+    validates :name,
+              :presence                => true,
+              :uniqueness_when_changed => [:set_type, :userid, :group_id],
+              :if                      => proc { |c| c.class.in_my_region.exists?(:name => c.name) }
+    validates :description,
+              :presence => true
 
-    belongs_to  :owner, :polymorphic => true
+    belongs_to :owner,
+               :polymorphic => true
+    has_many   :miq_set_memberships,
+               :foreign_key => :miq_set_id,
+               :dependent   => :delete_all
 
     acts_as_miq_taggable
 
-    alias_with_relationship_type :members,            :children
-    alias_with_relationship_type :miq_sets,           :children
-    alias_with_relationship_type :remove_member,      :remove_child
-    alias_with_relationship_type :remove_all_members, :remove_all_children
-    alias_with_relationship_type :add_members,        :add_children
-
-    alias_method model_table_name.to_sym, :children
-
-    virtual_has_many model_table_name.to_sym, :uses => :all_relationships
+    [:members, :miq_sets, :children, model_table_name.to_sym].each do |hm_relation|
+      has_many hm_relation,
+               :through     => :miq_set_memberships,
+               :source      => :member,
+               :source_type => model_class.name.to_s
+    end
```

Comment:
can this be done with rails?

```suggestion
    has_many :members,
             :through     => :miq_set_memberships,
             :source      => :member,
             :source_type => model_class.name.to_s
    alias_attribute :miq_sets, :members
    alias_attribute :children, :members,
    alias_attribute model_table_name.to_sym, :members
```

also not sure that the `to_sym` is necessary.

Reply:
Possibly.  I wasn't sure if that would maintain all of the other setters and such with just an `alias`, so I decided to define it manually.  Can give it a shot.

Reply:
Going to hold off on this one since I expect this has a good chance of having issues.  We circle and test it out once it is easier to test without cross repo.

Reply:
agreed.
sure does feel like we have so many redundant terms.

Reply:
Good points... I wonder if it's easier to find the callers and try changing them to the most common form... then if all callers can be changed, we can remove the unused method/attribute.